### ISSUE TITLE
Thoughts on hooks as arrays rather than strings?

### ISSUE THREAD

**Issue by victorquinn** [apex/apex#258]
Greetings, I've been using Apex for the past few days and loving it!

One issue we have is that our build hook is growing rather unwieldy.

We've got shared code we're copying into the lambda, we're using babel to compile our js code, copying in the node modules, and so on. So our build hook is now one big ass string with a lot of ampersands.

Perhaps this has already been discussed (and if so my bad!), but any opposition to changing the hooks from a single string to an array of strings and executing each in sequence?

I'd be happy to submit a PR but this would be a breaking change so I'd like to hear feedback before doing so.

**Comment by mthenw** [comment]
Hey @victorquinn,

maybe you could consider using Makefile or simple bash script?

The problem that I see with hook being array of string is what should happen if one of the scripts return error exit code. Should we stop, should we continue? There are always use cases that are exceptive. That's why using script/makefile/grunt etc is the best way.

**Comment by tj** [comment]
I thought about adding array support and just joining with `&&`, but I think in most cases you can split between deploy/build hooks to keep it short-ish. Run linting/tests in deploy and babel in build, similar to https://github.com/apex/apex/commit/4f4b6f8277cb34ee0b0ebc7b66a106ed6f6fccca which is a lot shorter now.

I usually use makefiles but I think what npm people usually do is use separate scripts and do `npm run foo && npm run bar`, but we don't support arbitrarily-named scripts yet ( #210).

**Comment by tj** [comment]
the ability to treat each hook as an array would be useful internally as well, prepend/append could be nice for plugins

**Comment by victorquinn** [comment]
FYI I actually originally tried using a Makefile but stumbled because the code in the hooks is run from the function directory rather than the root of the project.

So unless I added a Makefile in each function it wouldn't work by just running `make`. Eventually figured out I had to make the hook `make -f ../../Makefile`

**Comment by tj** [comment]
The only reason I'm hesitant to add it is because I can't think of many use-cases that end up very large. testing, linting, and building is about it but we've got two hooks that those can be spread across.

**Comment by victorquinn** [comment]
Sorry, should have clarified, I didn't make my last comment in order to make the case for the change, just adding that in case anyone else stumbled similarly. 

I'm mostly ambivalent, while it does make some things easier, it complicates others as you've noted.

**Comment by tj** [comment]
yep no worries :D just dumping my thoughts in case anyone else finds more use-cases that would benefit

**Comment by CyborgMaster** [comment]
I think it would be great to be able to have individual function.json files add to the hooks.  So I can define project wide hooks and then function specific ones.  An array would work for that, but I think the inheritance methods for the json still wouldn't work in this case.  It seems that arrays are replaced when inherited, not added to.
